lib/ansible/modules/cloud/cloudstack/cs_image_store.py

```diff
@@ -0,0 +1,203 @@
+    description: the protocol of the image store
+    type: str
+    returned: success
+provider_name:
+    description: the provider name of the image store
+    type: str
+    returned: success
+scope:
+    description: the scope of the image store
+    type: str
+    returned: success
+url:
+    description: the url of the image store
+    type: str
+    returned: success
+zone:
+    description: the Zone name of the image store
+    type: str
+    returned: success
+
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.cloudstack import AnsibleCloudStack, cs_argument_spec, cs_required_together
+
+
+class AnsibleCloudstackImageStore(AnsibleCloudStack):
+
+    def __init__(self, module):
+        super(AnsibleCloudstackImageStore, self).__init__(module)
+        self.returns = {
+            'id': 'id',
+            'name': 'name',
+            'protocol': 'protocol',
+            'providername': 'provider_name',
+            'scope': 'scope',
+            'url': 'url',
+            'zone': 'zone'
+
+        }
+        self.image_store = None
+
+    def get_storage_providers(self, storage_type="image"):
+        args = {
+            'type': storage_type
+        }
+        storage_provides = self.query_api('listStorageProviders', **args)
+        return [provider.get('name') for provider in storage_provides.get('dataStoreProvider')]
+
+    def get_image_store(self):
+        image_store = self.module.params.get('name')
+        args = {
+            'name': self.module.params.get('name'),
+            'zoneid': self.get_zone(key='id')
+        }
+
+        image_stores = self.query_api('listImageStores', **args)
+        if image_stores:
+            for img_s in image_stores.get('imagestore'):
+                if image_store.lower() in [img_s['name'].lower(), img_s['id']]:
+                    self.image_store = img_s
+                    break
+
+        return self.image_store
+
+    def present_image_store(self):
+        provider_list = self.get_storage_providers()
+        image_store = self.get_image_store()
+
+        if self.module.params.get('provider') not in provider_list:
+            self.module.fail_json(
+                msg='Provider %s is not in the provider list (%s). Please specify a correct provider' % (
+                    self.module.params.get('provider'), provider_list))
+
+        if not self.module.check_mode:
+            args = {
+                'name': self.module.params.get('name'),
+                'url': self.module.params.get('url'),
+                'zoneid': self.get_zone(key='id'),
+                'provider': self.module.params.get('provider')
+            }
+            if not image_store:
+                self.result['changed'] = True
+                self.query_api('addImageStore', **args)
+            else:
+                if self.has_changed(args, image_store):
+                    self.fail_json(msg='Image Store cannot be updated. Please delete it first.')
```

Comment:
I don't like to fail at this point. There are way better options to handle this, we could either implement a new force param to let a user "remove and add" an new image store by the users intention and if not force, just show a warning message to the users that we can not change the image store, but would recreate on force. We already have such a logic for in the cs_instance module for changing the offering on a running instance.

Reply:
Yeah, I agree that this could've been handled better. To be clear I did consider such an approach but was not sure if "recreate on update" is the right way. Adding a `force` param does seem like a good idea. I'll go ahead and implement it.

Reply:
This is done. @resmo can you please check?